### ISSUE TITLE
Logstash Mixing Throttled and Non-Throttled inputs

### ISSUE THREAD

**Issue by ayashjorden** [elastic/logstash#3642]
Hi,

I'm opening this ticket on the root Logstash repository as the title may be relevant to more than one of plugin.  Feel free to reassign as necessary.
## Intro

We use Logstash with both throttled and unthrottled inputs.
The configuration is as follows:
- Elasticsearch 1.7.0 2x6-core Intel Xeon 2.0GHZ with HT (total of 24 cores), 30GB heap, 10x1TB partitions.
  - `
    java version "1.7.0_79"
    OpenJDK Runtime Environment (IcedTea 2.5.5) (7u79-2.5.5-1~deb8u1)
    OpenJDK 64-Bit Server VM (build 24.79-b02, mixed mode)`
  - Cluster consists of one data+master node and a search-load-balancer node (master/data=false)
- Logstash 1.5.2, OpenVZ container, Debian 7, 8GB memory, 3GB heap, 64 filter threads.
  - ``````
    java version "1.7.0_75"
    OpenJDK Runtime Environment (IcedTea 2.5.4) (7u75-2.5.4-1~deb7u1)
    OpenJDK 64-Bit Server VM (build 24.75-b04, mixed mode)```
    ``````
  - Lumberjack input plugin is latest version and includes the circuit-breaker mechanism
  - **Configuration example:**

``` javascript
input {
    lumberjack {
        host            => "10.10.10.10"
        port            => 5230
        ssl_certificate => '/ssl/logstash-forwarder.crt'
        ssl_key         => '/ssl/logstash-forwarder.key'
    }
    tcp {
        port => 5514
        type => "CDN"
        ssl_cert => "/ssl/logstash-forwarder.crt"
        ssl_key => "/ssl/logstash-forwarder.key"
        ssl_enable => true
        codec => line {
            charset => "ISO-8859-1"
        }
    }
}
filter {
# Grok filters for Postfix, CDN, Nginx and application logs
# General structure is:
if [type] == “<SOME_TYPE>” {
# process message with grok and other filter types.
}
}
output {
    if [@metadata][type] in [ "postfix", "CDN"] {
        elasticsearch {
            host     => [ "10.10.0.20" ]
            protocol => "transport"
            port     => "9300"
            cluster  => "elasticsearch"
            index    => "%{[@metadata][type]}-%{+YYYY.MM.dd}"
            document_type => "%{[@metadata][type]}"
            manage_template => false
        }
    } 
**# 3 more elasticsearch outputs to handle the rest of the event types.**
    else {
        file {
            path => "/var/log/logstash/unknown_messages.log"
        }
    }
}
```
## Current Behaviour

In times of low throughput (outside heavy input hours) the event flow seems balanced between CDN logs, and _all_ other log types.
During heavy input hours, we suffer from **serious** degradation in lumberjack inputs (Nginx / Rails / Postfix), whereas TCP input (CDN log) throughput stays pretty much the same.

These images show the discrepancy between TCP logs and other (lumberjack) logs for a given time period:
TCP records:

![cdn_logs](https://cloud.githubusercontent.com/assets/4204371/8850102/df163ea2-314d-11e5-9362-35116433a0f5.png)

Lumberjack (Nginx) records:

![nginx_logs](https://cloud.githubusercontent.com/assets/4204371/8850123/f27ec50e-314d-11e5-92fc-93a2b412737d.png)

Nginx event count, by production host

![nginx_by_host](https://cloud.githubusercontent.com/assets/4204371/8850124/f7b453f4-314d-11e5-896b-02a7468c63ce.png)

NB from around 21:00, the TCP logs drop off completely for some time, and the lumberjack input is much lower from then on.  This is detailed in the last screenshot above, where each colour is a production app server (we have four high volume production app servers on lumberjack inputs, as well as our low-volume staging servers - each one outputs both nginx and rails via logstash-forwarder).  It’s interesting (and possibly relevant) that prior to the hiccough, all four servers were pretty much throwing an equal number of log information, but afterwards, the proportions seemed to change.

Below are the exceptions we got from Logstash log files during the high-throughput timeframe:

``` javascript
{:timestamp=>"2015-07-20T18:48:43.612000+0000", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2015-07-20T18:48:43.613000+0000", :message=>"Exception in lumberjack input thread", :exception=>#<LogStash::CircuitBreaker::OpenBreaker: for Lumberjack input>, :level=>:error}
{:timestamp=>"2015-07-20T18:48:43.666000+0000", :message=>"CircuitBreaker::rescuing exceptions", :name=>"Lumberjack input", :exception=>LogStash::SizedQueueTimeout::TimeoutError, :level=>:warn}
{:timestamp=>"2015-07-20T18:48:43.667000+0000", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2015-07-20T18:48:43.668000+0000", :message=>"Exception in lumberjack input thread", :exception=>#<LogStash::CircuitBreaker::OpenBreaker: for Lumberjack input>, :level=>:error}
{:timestamp=>"2015-07-20T18:48:43.729000+0000", :message=>"Lumberjack input: the pipeline is blocked, temporary refusing new connection.", :level=>:warn}
{:timestamp=>"2015-07-20T18:48:44.229000+0000", :message=>"Lumberjack input: the pipeline is blocked, temporary refusing new connection.", :level=>:warn}
{:timestamp=>"2015-07-20T18:48:44.730000+0000", :message=>"Lumberjack input: the pipeline is blocked, temporary refusing new connection.", :level=>:warn}
{:timestamp=>"2015-07-20T18:48:45.231000+0000", :message=>"Lumberjack input: the pipeline is blocked, temporary refusing new connection.", :level=>:warn}
{:timestamp=>"2015-07-20T18:48:45.251000+0000", :message=>"CircuitBreaker::rescuing exceptions", :name=>"Lumberjack input", :exception=>LogStash::SizedQueueTimeout::TimeoutError, :level=>:warn}
```

Here’s what we **assume** might be happening, based on the behaviour described above:
- The TCP input which has no circuit-breaker hammers the pipeline and blocks (or throttles) the lumberjack input from pushing events. When lumberjack input can’t push event to its filter event-queue, it refuses new connections. 
- The lumberjack circuit-breaker limits shippers one by one, starting with the heavy hitters. Logic is equivalent or at least similar to TCP - repeated cuts in half the throughput of the top abusers until things stabilize, and then a slow raising of limits.

_Out of all this, comes the following:_
1. Is there any basis for our assumptions above?
2. Is our use case sane?  Is such a setup (mixing circuit-breaker and non-circuit-breaker inputs) sane?
3. Can any logstash developers confirm or deny our hypotheses above?

_Bottom line:_
What do we have to do to enable our Logstash instance to cope with this?   We’re happy to answer any requests for more detail on our setup, and try any suggestions that the community may have about how to move forward.

Thanks,
Yarden and DevOps Team

**Comment by ayashjorden** [comment]
Hi,
I see there's no update on this ticket.

Can I provide any additional (maybe missing) data?

Thanks,
Yarden

**Comment by ayashjorden** [comment]
Hi @suyograo , 
Will this ticket/issue be more visible to the community if I'll move it to discuss.elastic.co ?

Thanks,
Yarden

**Comment by suyograo** [comment]
@ayashjorden we'll review it here soon, but I also encourage to cross post it on discuss.elastic.co

**Comment by ayashjorden** [comment]
Will do.

**Comment by ayashjorden** [comment]
HI again @suyograo ,

Hoped this issue will attractive for Logstash devs/contributors.

Any updates/thoughts/ideas? 

Thanks,
Yarden

**Comment by wfindley** [comment]
I'm seeing this behavior too.  It seems very like the reports in #3575

**Comment by ayashjorden** [comment]
Hi @wfindley,
Thanks for your comment.
I understand the similarity to #3575 (circuit-breaker kick in) but its subject is about Logstash not shutting down. (or I missed something...)

This issue aims to understand the proper use of Logstash's architecture by presenting a use-case which the community can discuss and reason about.

**Comment by antgel** [comment]
@suyograo Hi, we put it [here](https://discuss.elastic.co/t/logstash-mixing-throttled-and-non-throttled-inputs/26548), but for some reason it's generated no discussion.  It's really blocking us.  :(

**Comment by davewongillies** [comment]
Not that it accounts for much, but we're experiencing this same isue with logtash 1.5.x too.

**Comment by davewongillies** [comment]
So in the end, in our case the issue ended up being that our ES cluster that logstash was sending to was running with the G1GC collector, which ES has issues with and is not currently recommended for production use with ES. We switched back to the standard CMS collector and the issues have gone away.

**Comment by ayashjorden** [comment]
Thank you @davewongillies for your input :)

We're using ES as it comes our of the box, G1GC is not in use.

**Comment by jordansissel** [comment]
> both throttled and unthrottled inputs

I don't know this concept. What is a "throttled input" or "unthrottled input" ?

**Comment by ayashjorden** [comment]
We consider a 'throttled' input any input that blocks/push-back it's source
when LS pipeline is full (output and filter queues are full) and can't
accept new events. E.g. lumberjack input with pushback mechanism enabled
(thats the default)

An 'unthrottled' input will keep trying pushing messages to LS pipeline
regardless if its full of not. E.g. tcp input.

both throttled and unthrottled inputs

I don't know this concept. What is a "throttled input" or "unthrottled
input" ?

—
Reply to this email directly or view it on GitHub
<https://github.com/elastic/logstash/issues/3642#issuecomment-140485726>.

**Comment by jordansissel** [comment]
@ayashjorden all inputs except UDP do this. The TCP input will stop reading (because it gets stuck) if the filter queue is full.

**Comment by jordansissel** [comment]
Based on the holes in your graphs, it seems like you are letting logstash choose the time for the event. I'd recommend using the date filter to have Logstash learn the real time the event occurred. This would not necessarily solve the short-term stalls you see, but it would remove the gaps in time because time wouldn't be pausing periodically. :)

**Comment by jordansissel** [comment]
^^ is a recommendation to help you resolve part, but not all, of this situation.

I agree it's weird that tcp input continues to function normally while lumberjack does not, but I suspect this is something you can fix with the lumberjack configuration. My guess, for now, is that this is a configuration issue, and it may be best discussed on discuss.elastic.co - I'm not sure there's a bug just yet.

**Comment by ayashjorden** [comment]
Re: using date filter for letting Logstash learn event date, we're using the date filter from day 1.
The graphs in the OP are just one example of many showing this behaviour. 

@jordansissel We [tried](https://discuss.elastic.co/t/logstash-mixing-throttled-and-non-throttled-inputs/26548) discussing on discuss.elastic.co, but let's give it another try :)

**Comment by fizerkhan** [comment]
+1

**Comment by ayashjorden** [comment]
@fizerkhan join us on [discuss.elastic](https://discuss.elastic.co/t/logstash-mixing-throttled-and-non-throttled-inputs/26548)

**Comment by suyograo** [comment]
Closing this since there's an ongoing discussion on our fourms.
